datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -3305,6 +3305,12 @@ mod tests {
         );
         assert_eq!(simplify(expr.clone()), lit(true));
 
+        // 3.5 c1 NOT IN (1, 2, 3, 4) OR c1 NOT IN (4, 5, 6, 7) -> c1 != 4 (4 overlaps)
```

Comment:
💯 This case shocks me (again) on how powerful this rule is

Reply:
@jayzhan211  gets all the credit for implementing it in https://github.com/apache/arrow-datafusion/pull/8949 (🏆)